test/test_matplotlib.py

```diff
@@ -28,3 +28,35 @@ def test_pdf(selenium):
     selenium.run("import io")
     selenium.run("fd = io.BytesIO()")
     selenium.run("plt.savefig(fd, format='pdf')")
+
+
+def test_rendering(selenium):
+    selenium.load_package("matplotlib")
+    selenium.run("""
+    from js import XMLHttpRequest
+    from matplotlib import pyplot as plt
+    import numpy as np
+    import pyodide, io
+    t = np.arange(0.0, 2.0, 0.01)
+    s = 1 + np.sin(2 * np.pi * t)
+    plt.plot(t, s, linewidth=1.0, marker=11)
+    plt.plot(t, t)
+    plt.grid(True)
+    plt.show()
+    canvas_data = plt.gcf().canvas.get_pixel_data()
+    """)
+    selenium.run("""
+    def _get_url_async(url):
+        req = XMLHttpRequest.new()
+        req.open('GET', url, True)
+        req.responseType = 'blob'
+        req.send(None)
+        if req.readyState == 4:
+            return req.response
+        else:
+            return 'nay'
+
+    ref_data = _get_url_async('test/canvas.png')
+    """)
+    x = selenium.run("ref_data")
+    assert x == 1
```

Comment:
@mdboom I have tried numerous ways, and I ultimately settled to this. Here, the value of `x` is `nay`. If I do not return `nay`, it's `None`. Thus, it can be inferred that the code never goes to the condition `if req.readyState == 4:` which kinda means a not successful transfer. I suspect that the URL `test/canvas.png` doesn't work except for the function `pyodide.open_url`. This is because when I ran the test, I could see the status code as `500`. Any tips on how to get this right?

Reply:
Alternatively, I decided to use `fetch`, `Request` and `Response`

Usually, the code for `fetch` on JS is used as
```
var myImage = document.createElement('img');
var myRequest = new Request('https://i.ibb.co/3f4yJQS/face4.jpg');

fetch(myRequest)
.then(function(response) {
  return response.blob();
})
.then(function(myBlob) {
  var objectURL = URL.createObjectURL(myBlob);
  myImage.src = objectURL;
});

document.body.appendChild(myImage);
```

For the python side, I tried the following but got stuck since I do not know how to write promises on Python side.
```
from js import Request, fetch, Response
req = Request.new('test/canvas.png')
resp = fetch(req)
# how to convert resp to blob using Response?
```

Reply:
I have also tried the following snippet on an Iodide Notebook - in an attempt to use the `fetch` API from the Python side.

```
%%py

from js import document, Request, fetch, URL
img_tag = document.createElement('img')
req = Request.new('https://i.ibb.co/3f4yJQS/face4.jpg')

def func(response):
    return response.blob()

def func2(blob):
    objURL = URL.createObjectURL(myBlob)
    img_tag.src = objURL

fetch(req).then(func).then(func2)

document.body.appendChild(img_tag)
```

This yields the error `TypeError: 'fetch' called on an object that does not implement interface Window.` The same error comes on the terminal while running the test too.

Reply:
You can pass Python functions to `then`, just like in Javascript.  The only difference is that Python doesn't really have anonymous functions, so you have to explicitly define them.  So the example above converts to:

```
def get_blob(response):
    return response.blob()

def set_object_url(my_blob):
    object_url = URL.createObjectURL(my_blob)
    myImage.src = object_url

fetch(my_request).then(get_blob).then(set_object_url)
```